packages/jupyter-ai/jupyter_ai/chat_handlers/ask.py

```diff
@@ -67,6 +63,13 @@ def create_llm_chain(
             memory_key="chat_history", return_messages=True, k=2
         )
 
+        if self._retriever is None:
+            learn_chat_handler = self.chat_handlers.get("/learn")
+            if not isinstance(learn_chat_handler, LearnChatHandler):
+                raise CustomLearnException()
+
+            self._retriever = Retriever(learn_chat_handler=learn_chat_handler)
+
```

Comment:
The retriever only needs to be initialized once, but this initializes it every time the chat model changes. You can use a property to cache the instance after it is loaded, like this:

```py
def __init__(...):
    ...
    self._retriever = None

@property
def retriever(self) -> BaseRetriever:
    if self._retriever:
        return self._retriever
    ...
```

This should also change all other mentions of `self._retriever` to `self.retriever`.